In [28]:
import pandas as pd


In [29]:
def get_last_season_xGD(row):
    if row["Squad"] in df_previous_year["Squad"].values:
        previous_year_xGD = df_previous_year[df_previous_year["Squad"] == row["Squad"]]["xGD"].values[0]
        return previous_year_xGD * 0.2
    else:
        return 0

In [30]:
def get_last_season_xGD_2(row):
    if row["Squad"] in df_previous_year_2["Squad"].values:
        previous_year_xGD = df_previous_year_2[df_previous_year_2["Squad"] == row["Squad"]]["xGD"].values[0]
        return previous_year_xGD * 0.1
    else:
        return 0

In [31]:
def get_prediction(row):
    # if row["Squad"] == "Juventus":
    #     return row['Pts']/4 + row["xGD"]*1 + row["Previous_transformed_xGD"] + row["Previous_transformed_xGD_2"] + 10
    # else:
    #     return row['Pts']/4 + row["xGD"]*0.8 + row["Previous_transformed_xGD"] + row["Previous_transformed_xGD_2"]

    return row['Pts']/4 + row["xGD"]*0.8 + row["Previous_transformed_xGD"] + row["Previous_transformed_xGD_2"]

### Dane z 3 ostatnich sezonów Serie A i Premier League

### Model predykcyjny jest oparty o aktualną ilość punktów i różnicę bramek wsparty o różnicę bramek w poprzednich latach.

In [32]:
import pandas as pd

df = pd.read_csv("serie_a_22_23.csv")
df_previous_year = pd.read_csv("serie_a_21_22.csv")
df_previous_year_2 = pd.read_csv("serie_a_20_21.csv")

df["Previous_transformed_xGD"] = df.apply(get_last_season_xGD, axis=1)
df["Previous_transformed_xGD_2"] = df.apply(get_last_season_xGD_2, axis=1)
df["Prediction_score"] = df.apply(get_prediction, axis=1)
print(df.sort_values(by="Prediction_score", ascending=False, ignore_index=True)[['Rk', 'Squad','Prediction_score']].to_string())

    Rk          Squad  Prediction_score
0    1         Napoli             31.65
1    5          Inter             30.11
2    7           Roma             27.51
3    2          Milan             24.34
4    6       Atalanta             21.16
5    3       Juventus             17.27
6    4          Lazio             11.21
7   10     Fiorentina             10.11
8    8        Udinese              6.41
9    9         Torino              3.20
10  14          Monza              3.12
11  15       Sassuolo              1.52
12  11        Bologna             -1.40
13  16          Lecce             -2.25
14  20  Hellas Verona             -3.08
15  18      Cremonese             -6.81
16  13         Empoli             -9.51
17  12    Salernitana            -10.19
18  19      Sampdoria            -11.29
19  17         Spezia            -12.82


### Serie A wygra Napoli lub Inter. Na trzeciej pozycji skończy Roma.

In [33]:
import pandas as pd

df = pd.read_csv("premier_22_23.csv")
df_previous_year = pd.read_csv("premier_21_22.csv")
df_previous_year_2 = pd.read_csv("premier_20_21.csv")

df["Previous_transformed_xGD"] = df.apply(get_last_season_xGD, axis=1)
df["Previous_transformed_xGD_2"] = df.apply(get_last_season_xGD_2, axis=1)
df["Prediction_score"] = df.apply(get_prediction, axis=1)
print(df.sort_values(by="Prediction_score", ascending=False, ignore_index=True)[['Rk', 'Squad','Prediction_score']].to_string())

    Rk            Squad  Prediction_score
0    2  Manchester City             39.02
1    1          Arsenal             24.77
2    6        Liverpool             23.27
3    4        Tottenham             17.97
4    7         Brighton             11.93
5    8          Chelsea             11.85
6    3    Newcastle Utd             10.33
7    5   Manchester Utd              9.98
8   16         West Ham              6.65
9   10        Brentford              4.19
10  12      Aston Villa              2.06
11  11   Crystal Palace              0.45
12  15     Leeds United             -0.95
13  13   Leicester City             -1.79
14   9           Fulham             -2.54
15  18  Nott'ham Forest             -3.79
16  19      Southampton             -4.94
17  14      Bournemouth             -6.08
18  20           Wolves             -7.14
19  17          Everton             -7.63


### Zdecydowanie lige wygra Manchester City, Arsenal i Liverpool na 2.3 pozycji. Arsenal ma zaniżoną pozycje ze względu na słabe wyniki w porpzednich latach.